sql: mark all PARTITION BY columns as NOT NULL on CREATE

otan opened the pull request:
Release note (sql change): To match the behaviour of PRIMARY KEY
creations setting all relevant fields to NOT NULL, all PARTITION BY
/ REGIONAL BY ROW columns will also have their fields set to NOT NULL on
CREATE TABLE time.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/60379)
<!-- Reviewable:end -->

ajstorm reviewed (approved):
Looks great.  As part of testing this out, I noticed that this error string might not print exactly as we'd like.  Can we fix it as part of this PR?

https://sourcegraph.com/github.com/cockroachdb/cockroach/-/blob/pkg/sql/alter_table_locality.go#L267

`ERROR: cannot use column c for REGIONAL BY ROW table as it may contain NULL values%!(EXTRA string=crdb_internal_region)`

Other than that, <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60379#-:-MTBGQK15OyOcO1MdqcE:b-9qwavb)* status: :shipit: complete! 1 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

otan commented:
heh i fixed this in a separate PR that is yet to come.

but done!

bors r=ajstorm

otan commented:
bors r-

otan committed:
sql: mark all PARTITION BY columns as NOT NULL on CREATE

Release note (sql change): To match the behaviour of PRIMARY KEY
creations setting all relevant fields to NOT NULL, all PARTITION BY
/ REGIONAL BY ROW columns will also have their fields set to NOT NULL on
CREATE TABLE time.

otan commented:
bors r=ajstorm
